util/reggen/countermeasure.py

```diff
@@ -0,0 +1,80 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Dict, List
+
+from .lib import check_keys, check_str, check_list
+
+# TO BE DEFINED
+CM_ASSETS = [
+    'FOO'
+]
+
+# TO BE DEFINED
+CM_TYPES = [
+    'BAR'
+]
+
+
+class CounterMeasure:
+    """Object holding details for one countermeasure within an IP block."""
+
+    def __init__(self, asset: str, cm_type: str, desc: str):
+        self.asset = asset
+        self.cm_type = cm_type
+        self.desc = desc
+
+    @staticmethod
+    def from_raw(what: str, raw: object) -> 'CounterMeasure':
+        """
+        Create a CounterMeasure object from a dict.
+
+        The dict must have the keys 'name' and 'desc', where 'name' has
+        to follow the canonical countermeasure naming convention.
+        """
+        rd = check_keys(raw, what, ['name', 'desc'], [])
+
+        name = check_str(rd['name'], f'name field of {what}')
+        desc = check_str(rd['desc'], f'desc field of {what}')
+
+        try:
+            asset, cm_type = name.split('.')
+        except ValueError:
+            raise ValueError(
+                f'Invalid countermeasure format: {name} ({what}).')
+
+        if asset not in CM_ASSETS:
+            raise ValueError(
+                f'Invalid countermeasure format: {name} ({what}).')
+
+        if cm_type not in CM_TYPES:
+            raise ValueError(
+                f'Invalid countermeasure format: {name} ({what}).')
+        return CounterMeasure(asset, cm_type, desc)
+
+    @staticmethod
+    def from_raw_list(what: str,
+                      raw: object) -> List['CounterMeasure']:
+        """
+        Create a list of CounterMeasure objects from a list of dicts.
+
+        The dicts must have the keys 'name' and 'desc', where 'name' has
+        to follow the canonical countermeasure naming convention.
+        """
+        ret = []
+        for idx, entry in enumerate(check_list(raw, what)):
+            entry_what = f'entry {idx} of {what}'
+            cm = CounterMeasure.from_raw(entry_what, entry)
+            ret.append(cm)
+        return ret
+
+    def _asdict(self) -> Dict[str, str]:
```

Comment:
[Unable to reply back to earlier comment on this]. Seems like this needs to be a public API. Recommend `as_dict` instead. Also, looks like this could be a static method. Fine with doing this cleanup (across other classes as well) later.

Reply:
This is part of the hjson API (we didn't invent the name).

Reply:
This is a standalone class that does not extend from anything else. What am I missing?

Reply:
The hjson library has a `dumps` function which works by converting an arbitrary object to nested lists and dictionaries and then serializing to a string. It knows how to handle lists, dicts, integers, strings etc (basically passing them straight through). For other arbitrary types, it uses Python's `getattr` function to check whether the type has an `_asdict` method. If so, it uses it and recurses. If not, it spits out an error and gives up.